File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/VoenmehScheduleParser.kt
package ru.bgtu_voenmeh.zapara.data

import java.time.LocalDate
import java.time.LocalTime
import ru.bgtu_voenmeh.zapara.data.api.JsonFail
import ru.bgtu_voenmeh.zapara.data.api.JsonValue
import ru.bgtu_voenmeh.zapara.data.api.StrictJson
import ru.bgtu_voenmeh.zapara.data.api.obj

object VoenmehScheduleParser {
    fun parseMeta(json: String): VoenmehScheduleMeta {
        val obj = payloadObject(json)
        val hasData = (obj.fields["has_data"] as? JsonValue.Bool)?.value ?: false
        if (!hasData) throw IllegalStateException("Расписание пока не загружено на сайте университета")
        val title = str(obj, "period")
        val groups = strings(obj.fields["groups"])
        if (groups.isEmpty()) throw IllegalStateException("Сайт университета не отдал список групп")
        return VoenmehScheduleMeta(
            periodStart = periodStart(title),
            weekCount = 2,
            periodTitle = title.ifBlank { "Расписание" },
            groups = groups,
            updatedAt = str(obj, "updated_at").ifBlank { null }
        )
    }

    fun parseLessons(json: String, groupId: String): List<Lesson> {
        val obj = payloadObject(json)
        val rows = (obj.fields["lessons"] as? JsonValue.Arr)?.items.orEmpty()
        val indexPerParity = mutableMapOf<Int, Int>()
        val lessons = ArrayList<Lesson>(rows.size)
        for (item in rows) {
            val row = try { item.obj() } catch (_: JsonFail) { continue }
            val day = int(row, "day")
            if (day !in 1..7) continue
            val parity = when (str(row, "week").lowercase()) {
                "odd", "нечетная", "нечётная" -> 1
                "even", "четная", "чётная" -> 2
                else -> 0
            }
            val kind = str(row, "kind").trim()
            val subject = str(row, "subject").trim()
            val discRaw = when {
                kind.isNotEmpty() && subject.isNotEmpty() -> "$kind $subject"
                subject.isNotEmpty() -> subject
                else -> kind
            }
            val timeStart = padTime(str(row, "time"))
            val timeEnd = try {
                val te = LocalTime.parse(timeStart).plusMinutes(95)
                "%02d:%02d".format(te.hour, te.minute)
            } catch (_: Exception) { "" }
            val teachers = strings(row.fields["teachers"])
            val rooms = strings(row.fields["rooms"])
            val classroomRaw = if (rooms.isEmpty()) "" else rooms.joinToString("; ").let { if (it.endsWith(";")) it else "$it;" }
            val place = GroupParser.placeOf(classroomRaw)
            val idx = (indexPerParity[parity] ?: 0) + 1
            indexPerParity[parity] = idx
            lessons.add(
                Lesson(
                    groupId = groupId,
                    dayOfWeek = day,
                    parity = parity,
                    index = idx,
                    timeStart = timeStart,
                    timeEnd = timeEnd,
                    subjectRaw = discRaw,
                    subjectNormalized = Parity.normalizeSubject(discRaw),
                    teacherRaw = teachers.joinToString("; "),
                    roomRaw = place.first,
                    buildingRaw = place.second,
                    typeRaw = kind,
                    classroomRaw = classroomRaw
                )
            )
        }
        return lessons
    }

    fun assemble(meta: VoenmehScheduleMeta, loaded: List<Pair<String, List<Lesson>>>): ParsedSchedule {
        val byName = loaded.associate { it.first to it.second }
        val groups = meta.groups.map { name ->
            GroupInfo(id = name, name = name, url = VoenmehScheduleClient.ORIGIN)
        }
        val lessons = ArrayList<Lesson>()
        for (g in groups) lessons += byName[g.id].orEmpty()
        return ParsedSchedule(groups, lessons, meta.periodStart, meta.weekCount, meta.periodTitle)
    }

    internal fun periodStart(title: String): LocalDate {
        val years = Regex("""(20\d{2})""").findAll(title).map { it.groupValues[1].toInt() }.toList()
        val year = years.firstOrNull() ?: 2026
        val spring = title.contains("весен", ignoreCase = true)
        return if (spring) LocalDate.of(year, 2, 9) else LocalDate.of(year, 9, 1)
    }

    private fun padTime(raw: String): String {
        val m = Regex("""(\d{1,2}):(\d{2})""").find(raw.trim()) ?: return ""
        return "%02d:%02d".format(m.groupValues[1].toInt(), m.groupValues[2].toInt())
    }

    private fun strings(value: JsonValue?): List<String> {
        val arr = value as? JsonValue.Arr ?: return emptyList()
        return arr.items.mapNotNull { (it as? JsonValue.Str)?.value?.trim()?.takeIf { s -> s.isNotEmpty() } }
    }

    private fun str(obj: JsonValue.Obj, name: String): String =
        (obj.fields[name] as? JsonValue.Str)?.value.orEmpty()

    private fun int(obj: JsonValue.Obj, name: String): Int {
        val n = obj.fields[name] as? JsonValue.Num ?: return 0
        return n.raw.toIntOrNull() ?: 0
    }

    private fun payloadObject(json: String): JsonValue.Obj {
        val trimmed = json.trimStart('\uFEFF', ' ', '\n', '\r', '\t')
        if (trimmed.startsWith("<")) throw IllegalStateException(TimetablePayload.NOT_XML)
        return try {
            StrictJson.parse(trimmed).obj()
        } catch (_: JsonFail) {
            throw IllegalStateException(TimetablePayload.NOT_XML)
        }
    }
}

data class VoenmehScheduleMeta(
    val periodStart: LocalDate,
    val weekCount: Int,
    val periodTitle: String,
    val groups: List<String>,
    val updatedAt: String?
)

File: src/Vograph.Timetable/VoenmehScheduleParser.cs
using System.Globalization;
using System.Text.Json;
using System.Text.RegularExpressions;
using Vograph.Core.Models;
using Vograph.Core.Services;
using Group = Vograph.Core.Models.Group;

namespace Vograph.Timetable;

public sealed record VoenmehScheduleMeta(
    DateTime PeriodStart,
    int WeekCount,
    string PeriodTitle,
    IReadOnlyList<string> Groups,
    string? UpdatedAt);

public sealed record ParsedSchedule(
    IReadOnlyList<Group> Groups,
    IReadOnlyList<Lesson> Lessons,
    DateTime PeriodStart,
    int WeekCount,
    string PeriodTitle,
    string? UpdatedAt = null,
    IReadOnlyList<string>? FetchedGroupNames = null);

public static class VoenmehScheduleParser
{
    public static VoenmehScheduleMeta ParseMeta(string json)
    {
        var root = PayloadObject(json);
        if (!Bool(root, "has_data"))
            throw new InvalidOperationException("Расписание пока не загружено на сайте университета");
        var title = Str(root, "period");
        var groups = Strings(root, "groups");
        if (groups.Count == 0)
            throw new InvalidOperationException("Сайт университета не отдал список групп");
        return new VoenmehScheduleMeta(
            PeriodStart(title),
            2,
            string.IsNullOrWhiteSpace(title) ? "Расписание" : title,
            groups,
            BlankToNull(Str(root, "updated_at")));
    }

    public static List<Lesson> ParseLessons(string json, string groupId)
    {
        var root = PayloadObject(json);
        var parsed = new List<Lesson>();
        if (!root.TryGetProperty("lessons", out var rows) || rows.ValueKind != JsonValueKind.Array)
            return NumberByTime(parsed);
        foreach (var item in rows.EnumerateArray())
        {
            if (item.ValueKind != JsonValueKind.Object) continue;
            var day = Int(item, "day");
            if (day is < 1 or > 7) continue;
            var week = Str(item, "week").ToLowerInvariant();
            var parity = week switch
            {
                "odd" or "нечетная" or "нечётная" => 1,
                "even" or "четная" or "чётная" => 2,
                _ => 0
            };
            var kind = Str(item, "kind").Trim();
            var subject = Str(item, "subject").Trim();
            var discRaw = kind.Length > 0 && subject.Length > 0 ? $"{kind} {subject}"
                : subject.Length > 0 ? subject : kind;
            var timeStart = PadTime(Str(item, "time"));
            var timeEnd = "";
            if (TimeSpan.TryParse(timeStart, CultureInfo.InvariantCulture, out var ts))
                timeEnd = ts.Add(TimeSpan.FromMinutes(95)).ToString(@"hh\:mm");
            var teachers = Strings(item, "teachers");
            var rooms = Strings(item, "rooms");
            var classroomRaw = rooms.Count == 0 ? "" : string.Join("; ", rooms);
            if (classroomRaw.Length > 0 && !classroomRaw.EndsWith(';')) classroomRaw += ";";
            var (roomRaw, buildingRaw) = TimetableParser.PlaceOf(classroomRaw);
            parsed.Add(new Lesson
            {
                GroupId = groupId,
                DayOfWeek = day,
                Parity = parity,
                Index = 0,
                TimeStart = timeStart,
                TimeEnd = timeEnd,
                SubjectRaw = discRaw,
                SubjectNormalized = ParityService.NormalizeSubject(discRaw),
                TeacherRaw = string.Join("; ", teachers),
                RoomRaw = roomRaw,
                BuildingRaw = buildingRaw,
                TypeRaw = kind,
                ClassroomRaw = classroomRaw
            });
        }
        return NumberByTime(parsed);
    }

    public static ParsedSchedule Assemble(VoenmehScheduleMeta meta, IEnumerable<(string Name, List<Lesson> Lessons)> loaded)
    {
        var byName = loaded.ToDictionary(x => x.Name, x => x.Lessons, StringComparer.Ordinal);
        var groups = meta.Groups.Select(name => new Group
        {
            Id = name,
            Name = name,
            Url = VoenmehScheduleClient.Origin
        }).ToList();
        var lessons = new List<Lesson>();
        foreach (var g in groups)
            if (byName.TryGetValue(g.Id, out var list)) lessons.AddRange(list);
        return new ParsedSchedule(groups, lessons, meta.PeriodStart, meta.WeekCount, meta.PeriodTitle, meta.UpdatedAt,
            loaded.Select(x => x.Name).ToList());
    }

    internal static List<Lesson> NumberByTime(List<Lesson> lessons)
    {
        foreach (var group in lessons.GroupBy(l => (l.DayOfWeek, l.Parity)))
        {
            var ordered = group.OrderBy(l => l.TimeStart, StringComparer.Ordinal).ThenBy(l => l.SubjectRaw, StringComparer.Ordinal).ToList();
            for (var i = 0; i < ordered.Count; i++) ordered[i].Index = i + 1;
        }
        return lessons;
    }

    internal static DateTime PeriodStart(string title)
    {
        var years = Regex.Matches(title, @"20\d{2}").Select(m => int.Parse(m.Value, CultureInfo.InvariantCulture)).ToList();
        var spring = title.Contains("весен", StringComparison.OrdinalIgnoreCase);
        var year = spring ? years.LastOrDefault(2026) : years.FirstOrDefault(2026);
        return spring ? new DateTime(year, 2, 9) : new DateTime(year, 9, 1);
    }

    private static string PadTime(string raw)
    {
        var m = Regex.Match(raw.Trim(), @"(\d{1,2}):(\d{2})");
        if (!m.Success) return "";
        return $"{int.Parse(m.Groups[1].Value, CultureInfo.InvariantCulture):00}:{m.Groups[2].Value}";
    }

    private static JsonElement PayloadObject(string json)
    {
        var trimmed = json.TrimStart('\uFEFF', ' ', '\n', '\r', '\t');
        if (TimetableParser.IsHtml(trimmed) || trimmed.StartsWith('<'))
            throw new InvalidOperationException(TimetableParser.NotTimetable);
        try
        {
            using var doc = JsonDocument.Parse(trimmed);
            if (doc.RootElement.ValueKind != JsonValueKind.Object)
                throw new InvalidOperationException(TimetableParser.NotTimetable);
            return doc.RootElement.Clone();
        }
        catch (JsonException)
        {
            throw new InvalidOperationException(TimetableParser.NotTimetable);
        }
    }

    private static List<string> Strings(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var arr) || arr.ValueKind != JsonValueKind.Array) return [];
        return arr.EnumerateArray()
            .Where(i => i.ValueKind == JsonValueKind.String)
            .Select(i => i.GetString()?.Trim() ?? "")
            .Where(s => s.Length > 0)
            .ToList();
    }

    private static string Str(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var v)) return "";
        return v.ValueKind == JsonValueKind.String ? v.GetString() ?? "" : "";
    }

    private static int Int(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var v)) return 0;
        if (v.ValueKind == JsonValueKind.Number && v.TryGetInt32(out var n)) return n;
        if (v.ValueKind == JsonValueKind.String && int.TryParse(v.GetString(), CultureInfo.InvariantCulture, out n)) return n;
        return 0;
    }

    private static bool Bool(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var v)) return false;
        return v.ValueKind == JsonValueKind.True;
    }

    private static string? BlankToNull(string value) => string.IsNullOrWhiteSpace(value) ? null : value;
}
